test/dart/_leak_checker_test.dart

```diff
@@ -0,0 +1,248 @@
+
+    // Drop totals and check signal.
+    leakProvider.value = _SummaryValues.zero;
+    await Future.delayed(doublePeriod);
+    stdout.checkAndClear(_SummaryValues.zero);
+    devtools.checkAndClear(_SummaryValues.zero);
+
+    await Future.delayed(doublePeriod);
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+  });
+
+  test('Listener-only checker sends leaks to just listener.', () async {
+    // ignore: unused_local_variable
+    final checker = leakChecker(
+      hasDevtools: false,
+      hasStdout: false,
+      hasListener: true,
+      checkPeriodically: true,
+    );
+
+    // Make sure there is no leaks.
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+    expect(listened.store, isEmpty);
+
+    await Future.delayed(doublePeriod);
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+    expect(listened.store, isEmpty);
+
+    // Report leaks and make sure they signaled one time.
+    leakProvider.value = _SummaryValues.nonZero;
+    await Future.delayed(doublePeriod);
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+    listened.checkAndClear(_SummaryValues.nonZero);
+
+    await Future.delayed(doublePeriod);
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+    expect(listened.store, isEmpty);
+
+    // Report the same leak totals and make sure there is no signals.
+    leakProvider.value = _SummaryValues.nonZeroCopy;
+    await Future.delayed(doublePeriod);
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+    expect(listened.store, isEmpty);
+
+    // Drop totals and check signal.
+    leakProvider.value = _SummaryValues.zero;
+    await Future.delayed(doublePeriod);
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+    listened.checkAndClear(_SummaryValues.zero);
+
+    await Future.delayed(doublePeriod);
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+    expect(listened.store, isEmpty);
+  });
+
+  test('No-timer checker sends leaks when checked.', () async {
+    // ignore: unused_local_variable
+    final checker = leakChecker(
+      checkPeriodically: false,
+      hasDevtools: true,
+      hasStdout: true,
+      hasListener: true,
+    );
+
+    // Make sure there is no leaks.
+    await Future.delayed(doublePeriod);
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+    expect(listened.store, isEmpty);
+
+    // Report leaks and make sure did not signal.
+    leakProvider.value = _SummaryValues.nonZero;
+    await Future.delayed(doublePeriod);
+    expect(stdout.store, isEmpty);
+    expect(devtools.store, isEmpty);
+    expect(listened.store, isEmpty);
+
+    // Check leaks and make sure they signaled one time.
+    checker.checkLeaks();
+    stdout.checkAndClear(_SummaryValues.nonZero);
+    devtools.checkAndClear(_SummaryValues.nonZero);
+    listened.checkAndClear(_SummaryValues.nonZero);
```

Comment:
expectations inside helper methods can make things a bit less clear,
It might be better to have an optional getter to get the values and do the expectation inside the test.

Reply:
I am not sure I understood your suggestion. If I just pull the method content to the test, it will make the test much bigger and less readable.
Renamed the method for now to make it more understandable.

Reply:
For example for each of these it would be clearer to state what is being tested in the test, not hiding it inside the helper :)
 
 it is a bit of a nit pick, but generally this is a practicer that is good to shoot for
    ```
    expect(listened.store, hasLength(1));
    expect(listened.store.first.matches(summary), isTrue);
    listened.clear();
    ```

Reply:
it will make it harder to read. how about passing array instead of one item? it should make it clear that the method validates the content of the array.

Reply:
I like that compromise for sure!

Reply:
Applied in https://github.com/dart-lang/leak_tracker/pull/20